### PR TITLE
fix(platform): address CI report and seed cache review findings

### PR DESCRIPTION
**Pull request by ntindle**
### Daybreak-reviewed follow-up — 2026-09-10

Current head: `ca578975fffb99d9f41d06467dec41abf78ec109`. Independent Daybreak review found no remaining blocker in the final follow-up diff. This is local independent review, not a substitute for required GitHub approvals.

- Reconcile Cobertura root totals with actual class source lines, traverse nested E2E packages, exclude duplicate method-line entries, and accommodate reporter rate truncation (absolute tolerance 0.0001).
- Apply the same validator to E2E coverage, including helper/test path filters and the existing full-stack validator test/JUnit step.
- Use /bin/bash on POSIX CI and resolve Bash on Windows, failing if unavailable. Preserve byte stdin and Git Bash support.
- Validation: 88 CI-helper tests + 38 container CI-report tests pass, zero failures/errors/skips. Five actual prior-head artifacts (four frontend shards and E2E) pass. Ruff, explicit S607, Black, workflow YAML and repository pre-commit all pass.
- The new commit is unsigned under the existing locked-key limitation; no signing settings changed. Prior-head green Actions runs do not validate this new SHA. New hosted CI/review is pending.
- Overlap remains advisory on PRs as Nick explicitly requested. No separate CI-status, build/test/security or publication gate was weakened.

### Why / What / How

Follow-up to the CI-speed stack #14285, based on #14352. Addresses current review findings on #14280, #14281 and #14282 without rewriting the previously reviewed stack branches. This is CI validation and cache correctness work, not another speed claim.

### Changes 🏗️

- Validate frontend Cobertura semantics (coverage root, summary counts/rate, source classes and executable lines), not merely XML syntax. Add helper tests with a JUnit artifact in Frontend CI. No coverage percentage threshold is introduced.
- Broaden the E2E SQL cache key to the complete backend and shared-library trees plus database configuration and the workflow. Bump the cache namespace to v3 and test the dependency boundary. Backend/fixture changes now invalidate the seed dump.
- Accept empty or omitted Results only for the secret-only Trivy report with valid image metadata. The Debian vulnerability report still requires results; explicit nulls, malformed records, wrong image names, findings and failed scan actions still fail. Trivy v0.70.0 marks Results with omitempty: https://github.com/aquasecurity/trivy/blob/v0.70.0/pkg/types/report.go#L32 .
- Replace the XML sanitization regex flagged by CodeQL #347 with explicit XML 1.0 character ranges, covered by boundary/round-trip tests. No alert was manually dismissed.
- Make invalid/empty Docker FinishedAt conversion fail with a specific smoke-test diagnostic. Preserve the stock stop timeout and elapsed-time checks; no retry/sleep added.
- Prove the Playwright validator already catches missing tests, returns failure, and preserves the original semantic-failure report.

No product services, test retries, matrix entries, publication permissions, or overlap advisory/blocking policy changed. Nick confirmed on 2026-09-10 that PR overlap detection must remain advisory. Its existing PR-only continue-on-error is retained; build/test/security failures and the separate Check PR Status aggregator are unchanged. This follow-up is required before merging the affected lower layers as a complete stack; do not treat an unresolved lower-layer defect as fixed if this follow-up is omitted.

### Agents and large language models used

Codex main agent: model/version unknown. Independent reviewer: gpt-daybreak-blue-latest (Codex).

#### For code changes:

- [x] I have clearly listed my changes in the PR description
- [x] I have made a test plan
- [x] I have tested my changes according to the test plan:
  - [x] 83 CI-helper tests and 35 container CI-report tests pass, zero skips; JUnit evidence retained locally.
  - [x] Cache dependency and clean-secret-report regression fixtures reproduced failures before their fixes.
  - [x] New coverage validator accepts four existing frontend shard artifacts. These are historical compatibility samples, not current-head CI evidence.
  - [x] Targeted Ruff/Black checks, UTF-8 workflow YAML parsing, and Bash syntax validation.
  - [x] Repository pre-commit hooks pass with native Git Bash on PATH. Initial WSL-shell path failure was corrected; no hook was bypassed.
  - [ ] Hosted CI and current-head reviews for the latest follow-up (pending).

#### For configuration changes:

- [x] .env.default is already compatible
- [x] docker-compose.yml is already compatible
- [x] Workflow/cache changes are listed above

Prior benchmark/Fable approvals remain historical evidence for older SHAs, not approval or timing evidence for this follow-up.

Commit 0fce44821b912071fad5a77b6325914d666998f0 is unsigned because the local signing key was locked; signing configuration was not changed.

---

> [!NOTE]
> **Low Risk**
> Changes are limited to GitHub Actions scripts, validators, and cache keys; no application auth, data paths, or deployment logic are modified.
> 
> **Overview**
> This follow-up tightens **CI report validation and E2E seed cache correctness** without changing product runtime behavior.
> 
> **Frontend coverage:** Integration shards now run a dedicated **Cobertura validator** (`validate_cobertura.py`) that checks summary counts, line-rate bounds, and source classes/lines—not just XML parseability. The lint job runs validator unit tests and uploads JUnit for them.
> 
> **E2E cache:** The full-stack workflow bumps the seed dump cache to **v3** and widens `hashFiles` to backend, `autogpt_libs`, compose, and `.env.default`. A **contract test** asserts the cache key globs still cover known transitive seed inputs.
> 
> **Single-container CI:** Trivy verification accepts **empty or omitted `Results` only** on `trivy-secrets.json` when `Metadata.ImageID` is a valid `sha256:` digest (Trivy v0.70 `omitempty`); vulnerability reports still require non-empty results. The smoke script’s `assert_clean_stop` **fails fast** on invalid/empty `FinishedAt` parsing with a clear message. JUnit XML sanitization drops the CodeQL-flagged regex in favor of **explicit XML 1.0 character ranges**, with boundary tests.
> 
> **Tests:** Added/extended coverage for Playwright validator behavior on missing `tests`, Cobertura CLI edge cases, Trivy empty-results rules, and smoke stop timestamp handling.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0fce44821b912071fad5a77b6325914d666998f0. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: .github/scripts/validate_cobertura.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        raise ValueError(f"{element.tag}.{attribute} must be a non-negative integer")
+    return int(value)
+
+
+def validate_report(root: ET.Element) -> None:
+    if root.tag != "coverage":
+        raise ValueError("report is not Cobertura coverage XML")
+    valid = nonnegative_integer(root, "lines-valid")
+    covered = nonnegative_integer(root, "lines-covered")
+    if valid == 0 or covered > valid:
```
[the remark this answers is not in the document]
Comment (ntindle) [reply .github/scripts/validate_cobertura.py:20]: Fixed in ca578975fffb99d9f41d06467dec41abf78ec109 after independent Daybreak assessment and final diff review. The validator now traverses packages//class (the E2E exporter nests package elements), counts only each class's direct lines/line entries (method lines are duplicates), and requires exact root total/covered counts. Rate uses rel_tol=0, abs_tol=1e-4 to accept the actual reporter's four-decimal truncation and float representations. New mismatch/nested-package/rounding/method-line fixtures cover the behavior; the original mismatches reproduced before the fix. All13 validator tests and all5 downloaded real prior-head CI coverage artifacts pass. New-head Actions are pending.

File: .github/workflows/platform-fullstack-ci.yml [hunk 2]
```diff
@@ -83,6 +85,7 @@ jobs:
           mkdir -p "$RUNNER_TEMP/fullstack-validator-results"
           poetry run pytest \
             ../../.github/scripts/test_validate_playwright_json.py \
+            ../../.github/scripts/test_e2e_seed_cache_contract.py \
```
[the remark this answers is not in the document]
Comment (ntindle) [reply .github/workflows/platform-fullstack-ci.yml:93]: Fixed in ca578975fffb99d9f41d06467dec41abf78ec109 after Daybreak review. E2E coverage now invokes the shared Cobertura validator at the unchanged coverage/e2e/cobertura-coverage.xml path. Added the helper and tests to both push/PR path filters and the existing full-stack validator pytest/JUnit step. Preserved always() validation, report uploads, Codecov and thresholds. Nested E2E package handling was added before wiring this: the real prior-head E2E report reconciles exactly (23630 source lines,7777 covered) and passes the stricter validator. New-head hosted CI remains pending.

File: autogpt_platform/single-container/tests/test_ci_smoke_stop.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+                    "STOCK_DOCKER_STOP_TIMEOUT=10\n"
+                    "CONTAINER_NAME=test-container\n"
+                    "docker() { printf '%s' invalid-timestamp; }\n"
+                    f"date() {{ return {date_status}; }}\n"
+                    "awk() { echo UNEXPECTED_AWK >&2; return 1; }\n"
+                    f"{function}\n"
+                    "assert_clean_stop test-stop\n"
+                )
+                result = subprocess.run(
+                    ["bash", "--noprofile", "--norc", "-s"],
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/single-container/tests/test_ci_smoke_stop.py]: Addressed in ca578975fffb99d9f41d06467dec41abf78ec109 using Daybreak's portable recommendation: /bin/bash on POSIX/Ubuntu CI; on Windows use shutil.which('bash') and fail clearly if unavailable. An unconditional /bin/bash would break native Windows Python/Git Bash, so that exact suggestion was not used. The subprocess receives the selected executable path, retains -s and byte stdin, and still uses only fixed test-controlled content. Added tests for POSIX selection, Windows resolution and missing-tool failure. All4 smoke-stop tests pass with native Git Bash; Ruff S607 and repository pre-commit pass. Daybreak classified this as a low-risk test lint/portability issue, not a production security defect.

File: .github/workflows/platform-fullstack-ci.yml [hunk 4]
```diff
@@ -384,13 +391,9 @@ jobs:
...
-            echo "missing machine-readable report: $report" >&2
-            exit 1
-          fi
-          python -c "import xml.etree.ElementTree as ET; ET.parse('$report')"
+        run: >-
+          python ../../.github/scripts/validate_cobertura.py
+          coverage/e2e/cobertura-coverage.xml
 
       - name: Upload E2E coverage to Codecov
         if: ${{ !cancelled() }}
```
[the remark this answers is not in the document]
Comment (ntindle) [reply .github/workflows/platform-fullstack-ci.yml:399]: Confirmed. Fixed in the owning workflow PRs during the requested restack: #14281 at 21a2c6a1893867637a2f2f9df35f44323d6102d2 gives the E2E validator an id and requires its successful outcome before Codecov upload. The equivalent frontend gap is fixed in #14280 at 0899b421d0a7431757a4af3982f4562c06aa8621. Both new workflow-contract tests failed before the fixes and pass afterward; rejected reports still upload as failure artifacts. The final integrated helper suites pass 90 + 38 tests. #14503 is superseded by the workflow-specific PRs, with its branch/history retained.
